_posts/2018-06-18-On-Obsessive-Statelessness.markdown

```diff
@@ -0,0 +1,201 @@
+---
+layout: epic
+title: "On the obsessive statelessness of Peril"
+date: 2018-06-18
+author: [orta]
+categories: [danger, peril, state]
+css: danger
+---
+
+We're at 9 months of serious usage of Peril in Artsy. However, I've been worried.
+
+To get you up to speed on Peril, Peril is a tool that takes GitHub webhooks, and makes it easy to build one-off
+actions. It does this by having a per-account settings JSON, that connects JavaScript files to events from webhooks.
+So, for example, you can write a rule which runs when closing an issue in GitHub that looks for associated Jira
+tickets and resolves them. Peril provides no implicit actions like that, it instead offers a JavaScript runtime
+environment optimised to this domain so you can make actions to fit your needs. Like a collection of single-file
+[probots][probots].
+
+Three months ago I started building out a "true" staging environment for Peril, one that allows any user or org on
+GitHub to click a button and have Peril running on their account. Pulling this off has two real interesting
+problems. Problem number one, security. Problem number two, my wallet.
+
+Both of these issues stem from one simple problem: I need to run other people's code on my machines and I think they
+should be able to store data. Which to be quite frank, is horrifying for a side-project. So, this post explores one
+of main aspects which I've architected Peril to make this problem tractable. Avoiding storing state in the form of
+data.
+
+<!-- more -->
+
+## Evaluation Context
+
+Let's start with grounding how Peril works. The GitHub term for when someone adds Peril to their account is that it
+creates an "Installation" of the
+[GitHub App](https://blog.github.com/2016-09-14-a-whole-new-github-universe-announcing-new-tools-forums-and-features/#integrate-seamlessly-with-github).
+When a webhook from GitHub is sent to Peril, Peril grabs the installation ID out a mongo database (yeah, I know,
+[how early-2010s][nosqliscool]) and pulls out a set of rules. These rules are a map of Webhook events and actions to
+files. For example:
+
+```json
+{
+  "rules": {
+    "issues.opened": "artsy/peril-settings@danger/new-rfc.ts",
+    "issue_comment": "artsy/peril-settings@org/markAsMergeOnGreen.ts",
+    "pull_request.closed": "artsy/peril-settings@org/closed-prs.ts"
```

Comment:
My first thought here is curiosity about the path scheme. Are these relative URLs, file paths, or a more abstract namespace (like JS imports)? This might be answered below, but I'm commenting as I read through. I can suspend understanding, but that's a hanging question in my mind as I read.

Reply:
I've added a link for people to dive into that if they care ( and if you want it anyway,  https://github.com/danger/peril/blob/93439da3a088e9a7824192e24d33295ced017239/docs/settings_repo_info.md )